andela/cyclops-backend
# #167728028 User role settings

**What does this PR do?**
- Verifies a user account through email
- Creates a role and permission based access control for the entire api

**Description of Task to be completed?**
Have the following endpoints working
- GET `/api/v1/users`
- GET `/api/v1/users/:email`
- GET `/api/v1/auth/confirm-email`
- PUT `/api/v1/admin/assign-role`
- PUT `/api/v1/admin/assign-permission`

**How should this be manually tested?**
   - Fetch the branch and run `npm i` on your terminal
      1. Account verification
         - Create an API key on [sendGrid](www.sendgrid.com) and copy it
         - In your `.env` file, set `SENDGRID_API_KEY` to hold the key
         - Run `npm run dev` and jump to postman
         - Go to the signup endpoint and signup after which an account verification email will be sent to you.
         - Click on the link to verify your account
         - Go to the sign in route and sign in
         - Copy the token returned in the response body
         - Set the Authorization header of every protected route to hold it as such:  `Bearer token`
      2. User Role Settings
         - Sign in as a Super Administrator
         - Get users and copy the email of the user you want to assign new role to
         - Go to the endpoint for assigning role
         - Assign a new role to the user by passing the user's email and the role in the request body
         - Assign permissions to that role by sending the role and the permission in the request body
     I also created middle-wares to verify the roles and permissions of any logged in user accessing an authenticated route

**Any background context you want to provide?**
1. Available roles 

   - Super Administrator
   - Travel Administrator
   - Travel Team Member
   - Manager
   - Requester

2. Available permissions 

   - create
   - read
   - update
   - delete

**What are the relevant stories?**
- [#167728028](https://www.pivotaltracker.com/n/projects/2381984/stories/167728028)
- [#167728016](https://www.pivotaltracker.com/n/projects/2381984/stories/167728016)

**Screenshots (if appropriate)**
![Screenshot (92)](https://user-images.githubusercontent.com/29434750/64299918-e5cab800-cf72-11e9-9860-0d8974e82052.png)

![Screenshot (93)](https://user-images.githubusercontent.com/29434750/64386043-03118c00-d030-11e9-917a-a41ff3bda61c.png)
![Screenshot (94)](https://user-images.githubusercontent.com/29434750/64386106-2c321c80-d030-11e9-8334-55b0d4bba678.png)


**Questions:**
- None
